src/v/storage/disk_log_impl.cc

```diff
@@ -1688,6 +1688,11 @@ int64_t compaction_backlog_term(
   std::vector<ss::lw_shared_ptr<segment>> segs, double cf) {
     int64_t backlog = 0;
 
+    // Only compare each segment to a limited number of other segments, to
+    // avoid the loop below blowing up in runtime when there are many segments
+    // in the same term.
+    static constexpr size_t limit_lookahead = 8;
```

Comment:
nit: would be nice to update the comment for `compaction_backlog` to describe how we approximate the formula in there.